### ISSUE TITLE
Shared block: focus loss because the whole component is re-rendered when editing / saving

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#6204]
This is a good example of how some development patterns in use in modern JavaScript frameworks should be avoided for accessibility reasons. While it's tempting to re-render a whole UI section because it's just convenient from a development perspective, it's very likely that doing so can produce a focus loss.

When changing a block to a shared block, an UI with an "Edit" button gets displayed:

<img width="704" alt="screen shot 2018-04-16 at 13 30 01" src="https://user-images.githubusercontent.com/1682452/38806700-681baf86-417a-11e8-932c-52bfc24d9dae.png">
<img width="712" alt="screen shot 2018-04-16 at 13 30 21" src="https://user-images.githubusercontent.com/1682452/38806701-683d1716-417a-11e8-957a-d53d49c28b76.png">

Pressing "Edit" re-renders the whole component, so the "Edit" button is not present in the DOM any longer. Same happens when pressing Save or Cancel. Since the button was the focused element, this produces a focus loss for keyboard users. The focus loss is more evident in some browsers, for example IE11, where users are forced to start again keyboard navigation from the document root. Other browsers are a bit more smart and try to keep focus "in place". Regardless, technically there's a focus loss and assistive technologies may get very confused.

In this cases, there's the need to manage focus programmatically:
- when pressing Edit: focus should be moved to the input field
- when pressing Save or Cancel and the form closes: focus should be moved back to the invoking context: the Edit button

See also the React documentation notes about focus management:
https://reactjs.org/docs/accessibility.html#programmatically-managing-focus
